src/lib.rs

```diff
@@ -168,6 +168,39 @@ impl IndexedMesh {
             Ok(())
         }
     }
+
+    /// Calculates the volume of the mesh. In order for a correct volume calculation the mesh must
+    /// be valid. Use `validate` to make sure that the mesh doesn't contain any holes before
+    /// calculating the volume.
+    pub fn volume(&self) -> f32 {
```

Comment:
I think this method should return an Optional[f32] or a Result[f32, SomeErrorType], and include a call to validate.
Furthermore, for this to be the true volume, I think we'd have to validate that the winding number of all triangles is the same, no?

Reply:
> I think this method should return an Optional[f32] or a Result[f32, SomeErrorType], and include a call to validate.
I think I would switch it around; have the `volume > 0` check in the validate function. but it is your call.

>Furthermore, for this to be the true volume, I think we'd have to validate that the winding number of all triangles is the same, no?
I believe that the connected indices check in the validate function ensures that the winding order is the same for all tris.

Reply:
> I believe that the connected indices check in the validate function ensures that the winding order is the same for all tris.
Good point. I agree.

> I think I would switch it around; have the volume > 0 check in the validate function. but it is your call. (if you know your > mesh to be valid it is annoying if a library spends computation on validation)
How about this:
* rename validate() into is_closed_and_has_correct_winding().
* call is_closed_and_has_correct_winding() inside volume().
* create validate() as volume() > 0

Reply:
In my opinion it is bad practice to combine these kinds of validate checks inside "computation functions". It is better to execute such validation checks only once during or just after the construction of the data type. For instance coding a ray tracer; the direction of the ray always needs to be a unit length vector, but i'm not going to normalise the vector every time I compute something. Instead the normalisation is executed once when constructing the ray.

I would propose something like this.
```rust
impl IndexedMesh {
    ...
    pub fn validate(&self) -> Result<()> {
        self.is_closed()?;
        self.non_zero_tris()?;
        if self.volume() > 0.0 {
            Err(::std::io::Error::new(
                ::std::io::ErrorKind::InvalidData,
                "mesh has a negative volume",
            ))
        } else {
            Ok(())
        }
    }

    fn is_closed() -> Result<()> { ... }
    fn non_zero_tris() -> Result<()> { ... }
}
```
What you think about this? downside with my approach compared to yours is that if a user wants to know if a model is valid and want to know it's volume the volume is essentially computed twice.

Reply:
I think it is questionable to assume the one winding is the correct and thus the other winding is invalid.
Meaning even a mesh that has negative winding might be valid.

The thing about volume is, that the calculation is only correct, if the mesh is closed and has correct winding.
But the above volume() function will also return a volume for meshes that are not closed or have no correct winding. In my opinion this would be a bug.
If you really wouldn't want to include the other validation checks in volume(), then volume would have to renamed to:
fn volume_assuming_the_mesh_is_closed_and_has_clockwise_winding()